components/Activity/QuestionnaireTable.tsx

```diff
@@ -0,0 +1,84 @@
+import { observer } from 'mobx-react';
+import { FC } from 'react';
+import { Button, Container, Table } from 'react-bootstrap';
+
+import { Question } from '../../models/Question';
+import { i18n } from '../../models/Translation';
+
+const { t } = i18n;
+
+export interface QuestionnaireTableProps {
+  questionnaire: Question[];
+  onDelete: (id: string) => void;
+  onMove: (id: string, direction: number) => void;
+}
+
+export const QuestionnaireTable: FC<QuestionnaireTableProps> = observer(
+  ({ questionnaire, onDelete, onMove }) => {
+    const typeMap = {
+      text: t('text'),
+      url: t('link'),
+      '--': '--',
+    };
+
+    return (
+      <Container fluid>
+        <Table responsive className="my-3">
+          <thead>
+            <tr>
+              <th>#</th>
+              <th>{t('question_id')}</th>
+              <th>{t('question_description')}</th>
+              <th>{t('question_type')}</th>
+              <th>{t('option_config')}</th>
+              <th>{t('whether_multiple')}</th>
+              <th>{t('whether_required')}</th>
+              <th>{t('operation')}</th>
+            </tr>
+          </thead>
+          <tbody>
+            {questionnaire.map(
+              (
+                { id, title, options, multiple, type, required },
+                index,
+                { length },
+              ) => (
+                <tr key={id}>
+                  <td>{index + 1}</td>
+                  <td>{id}</td>
+                  <td>{title}</td>
+                  <td>{typeMap[type || '--']}</td>
```

Comment:
```suggestion
                  <td>{typeMap[type] || '--'}</td>
```
对象取不到值自然就回退到 `||` 后面的值了。

**表达式求值**和**视图显示**是两个层面的事，混为一谈会导致思路不清、代码复杂。

Reply:
好的，但这里type类型是个可选枚举, “undefined”不能作为索引类型使用，需要先判断type存在

Reply:
> 好的，但这里type类型是个可选枚举, “undefined”不能作为索引类型使用，需要先判断type存在

那可以在解构声明时给一个默认值

Reply:
啊，怎么没想到，好的